### ISSUE TITLE
Go-lang wrapper for nickel

### ISSUE THREAD

**Issue by olorin37** [tweag/nickel#1929]
# Problem

Multiple or almost most of DevOps tools are written in go, let's name `terraform` for instance. In its case a provider for reading nickel files to native terraform structures could be written, although providers can be written in any language only `go` is officially supported. Regarding it, I think the best is to stick to `go`, but to evaluate nickel with from `go` it would be required to have nickel library exporting its capabilities to C ABI and then it would be possible to execute it with using `cgo`. In scope of this project such go package could be prepared similarly to pyckel. (I know about opposite approach where terraform code is defined in nickel and then exported to json which terraform loads, I mean something different). And mind, terraform is just an example. 

# Solution
This project could contain go package implementation which would import nickel library with `cgo` usage, in turn rust code needs to be exported to library with C ABI. 

`cgo` is considered as not performant, but it for described use case it will be absolutely enough.

**Comment by yannham** [comment]
I agree that a C-compatible facade would be useful in general. The first step is to have first a Rust facade,  because as discussed in various channels, `nickel-lang-core` is mostly an internal library provided "as it is" and isn't very user-friendly (although it's not too bad either). So how I see it, we should:

1. Create a `nickel-lang-rs` library which asbtract the details of `nickel-lang-core`, exposing a smaller, simpler, and more stable API
2. Just make a C compatible wrapper out of 1.
3. Make a Go wrapper (or really any language out there with C FFI) out of 2.

**Comment by yannham** [comment]
The (to be published today) 1.15 release officially features a public Nickel API (`nickel-lang` crate), a C API, and corresponding go bindings at https://github.com/nickel-lang/go-nickel. Basically, the plan laid out in the previous message has been executed.
